event-dispatcher/src/main/java/io/sk8s/event/dispatcher/EventDispatchingHandler.java

```diff
@@ -16,55 +16,264 @@
+	@PreDestroy
+	public synchronized void tearDown() {
+		running = false;
+		this.notify();
+		kafkaConsumerMonitor.close();
+	}
+
+	private void updateWakeUpInterval() {
+		this.scalingFrequencyMs = this.functions.isEmpty() ? 0 : 100;
+	}
+
+	private class ScalingMonitor implements Runnable {
+
+		@Override
+		public void run() {
+			while (running) {
+				synchronized (EventDispatchingHandler.this) {
+
+					Map<KafkaConsumerMonitor.TopicGroup, List<KafkaConsumerMonitor.Offsets>> offsets = kafkaConsumerMonitor
+							.compute();
+					logOffsets(offsets);
+					Map<String, Long> lags = offsets.entrySet().stream()
+							.collect(Collectors.toMap(
+									e -> e.getKey().group,
+									e -> e.getValue().stream()
+											.mapToLong(KafkaConsumerMonitor.Offsets::getLag)
+											.max().getAsLong()));
+					functions.values().stream().forEach(
+							f -> {
+								int desired = computeDesiredNbReplicas(lags, f);
+								String name = f.getMetadata().getName();
+								Integer currentDesired = actualNbReplicas.computeIfAbsent(name, k -> desired);
+								System.out.println("Want " + desired + " for " + name);
+								if (desired < currentDesired) {
+									if (edgeInfos.computeIfAbsent(name,
+											n -> System.currentTimeMillis()) < System.currentTimeMillis()
+													- DOWN_EDGE_LINGER_PERIOD) {
+										deployer.deploy(f, desired);
+										actualNbReplicas.put(name, desired);
+										edgeInfos.remove(name);
+									}
+									else {
+										System.out.printf("Waiting another %dms before scaling down to %d for %s%n",
+												DOWN_EDGE_LINGER_PERIOD
+														- (System.currentTimeMillis() - edgeInfos.get(name)),
+												desired, name);
+									}
+								}
+								else if (desired > currentDesired) {
+									deployer.deploy(f, desired);
+									actualNbReplicas.put(name, desired);
+									edgeInfos.remove(name);
+								}
+								else {
+									edgeInfos.remove(name);
+								}
+							});
+
+					try {
+						EventDispatchingHandler.this.wait(scalingFrequencyMs);
+					}
+					catch (InterruptedException e) {
+						Thread.currentThread().interrupt();
+					}
+				}
+			}
+		}
+
+		/**
+		 * Compute the desired nb of replicas for a function. Each function defines 4 values:
+		 * <ul>
+		 * <li>minReplicas (>= 0, default 0)</li>
+		 * <li>maxReplicas (minReplicas <= maxReplicas <= nbPartitions, default nbPartitions)</li>
+		 * <li>l, the amount of lag required to trigger the first pod to appear, default 1</li>
+		 * <li>L, the amount of lag required to trigger all (maxReplicas) pods to appear. Default
+		 * 10.</li>
+		 * </ul>
+		 * This method linearly interpolates based on witnessed lag and clamps the result between
+		 * min/maxReplicas.
+		 */
+		private int computeDesiredNbReplicas(Map<String, Long> lags, XFunction f) {
+			String fnName = f.getMetadata().getName();
+			long lag = lags.get(fnName);
+
+			// TODO: those 3 numbers part of Function spec?
+			int bigL = 10;
```

Comment:
we should name these more descriptively IMO ("initialLag" and "maxLag"?... or something more accurate if necessary)